apps/cmp.c

```diff
@@ -0,0 +1,3256 @@
+            return 0;
+        }
+        GENERAL_NAME_free(n);
+    }
+    return 1;
+}
+
+/* TODO potentially move to apps/lib/apps.c */
+/*
+ * create cert store structure with certificates read from given file(s)
+ * returns pointer to created X509_STORE on success, NULL on error
+ */
+static X509_STORE *load_certstore(char *input, const char *desc)
+{
+    X509_STORE *store = NULL;
+    STACK_OF(X509) *certs = NULL;
+
+    if (input == NULL)
+        goto err;
+
+    while (input != NULL) {
+        char *next = next_item(input);           \
+
+        if (!load_certs_autofmt(input, &certs, 1, opt_otherpass, desc)
+                || !(store = sk_X509_to_store(store, certs))) {
+            /* CMP_err("out of memory"); */
+            X509_STORE_free(store);
+            store = NULL;
+            goto err;
+        }
+        sk_X509_pop_free(certs, X509_free);
+        certs = NULL;
+        input = next;
+    }
+ err:
+    sk_X509_pop_free(certs, X509_free);
+    return store;
+}
+
+/* TODO potentially move to apps/lib/apps.c */
+static STACK_OF(X509) *load_certs_multifile(char *files,
+                                            const char *pass, const char *desc)
+{
+    STACK_OF(X509) *certs = NULL;
+    STACK_OF(X509) *result = sk_X509_new_null();
+
+    if (files == NULL)
+        goto err;
+    if (result == NULL)
+        goto oom;
+
+    while (files != NULL) {
+        char *next = next_item(files);
+
+        if (!load_certs_autofmt(files, &certs, 0, pass, desc))
+            goto err;
+        if (!sk_X509_add1_certs(result, certs, 0, 1 /* no dups */, 0))
+            goto oom;
+        sk_X509_pop_free(certs, X509_free);
+        certs = NULL;
+        files = next;
+    }
+    return result;
+
+ oom:
+    BIO_printf(bio_err, "out of memory\n");
+ err:
+    sk_X509_pop_free(certs, X509_free);
+    sk_X509_pop_free(result, X509_free);
+    return NULL;
+}
+
+typedef int (*add_X509_stack_fn_t)(void *ctx, const STACK_OF(X509) *certs);
+typedef int (*add_X509_fn_t)(void *ctx, const X509 *cert);
+
+static int setup_certs(char *files, const char *desc, void *ctx,
+                       add_X509_stack_fn_t addn_fn, add_X509_fn_t add1_fn)
+{
+    int ret = 1;
+
+    if (files != NULL) {
+        STACK_OF(X509) *certs = load_certs_multifile(files, opt_otherpass,
+                                                     desc);
+        if (certs == NULL) {
+            ret = 0;
+        } else {
+            if (addn_fn != NULL) {
+                ret = (*addn_fn)(ctx, certs);
+            } else {
+                int i;
+
+                for (i = 0; i < sk_X509_num(certs /* may be NULL */); i++)
+                    ret &= (*add1_fn)(ctx, sk_X509_value(certs, i));
+            }
+            sk_X509_pop_free(certs, X509_free);
+        }
+    }
+    return ret;
+}
+
+
+/*
+ * parse and transform some options, checking their syntax.
+ * Returns 1 on success, 0 on error
+ */
+static int transform_opts(void)
```

Comment:
suggestions, and them change this method to return CMD_ERR or any CMP_ command code.
`opt_cmd_s` should be passed as an argument.

Reply:
Well, the `transform_opts()` function is not only about the `cmd` option,
and selecting the latter as the primary result could be a bit misleading in my view.